### PR TITLE
Bootstrapable client

### PR DESCRIPTION
**Pull request by sbernard31**
The pull request aim to make the Leshan Client bootstrapable.
The first step was to get the registration information from Lwm2m security/servers objects.
(This is a work in progress, do not merge it :p)

### CODE REVIEW

**Comment by sbernard31** [comment]
There is still some limitation in this version, but I think we are near to be able to integrate this in master. (after some review)

Here is a non exhaustive list of knowing issues :
- we support only 1 device mangement server
- we support only PSK (RPK and X509 need some modification in scandium to be implemented)
- I think we're not really clean about threadsafeness :)
- bootstrap sequence is not safe (we delete the bootstrap server if this is asked)
- Ugly update strategy  (lifetime -1)
- manage a client stop pending bootstrap session
- we accept request of unknown server

File: leshan-client-cf/src/main/java/org/eclipse/leshan/client/californium/LeshanClient.java [hunk 1]
```diff
@@ -18,114 +18,139 @@
...
-    private final AtomicBoolean clientServerStarted = new AtomicBoolean(false);
     private final CaliforniumLwM2mClientRequestSender requestSender;
-    private final List<LwM2mObjectEnabler> objectEnablers;
+    private final RegistrationEngine engine;
+    private final BootstrapHandler bootstrapHandler;
 
-    public LeshanClient(final InetSocketAddress serverAddress, final List<LwM2mObjectEnabler> objectEnablers) {
-        this(new InetSocketAddress("0", 0), serverAddress, new CoapServer(), objectEnablers);
-    }
+    public LeshanClient(final String endpoint, final InetSocketAddress localAddress,
```
Comment (jvermillard) [inline leshan-client-cf/src/main/java/org/eclipse/leshan/client/californium/LeshanClient.java]: javadoc?

File: leshan-client-cf/src/main/java/org/eclipse/leshan/client/californium/LeshanClientBuilder.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+    /**
+     * Sets the list of objects enablers
+     */
+    public LeshanClientBuilder setObjects(List<LwM2mObjectEnabler> objectEnablers) {
+        this.objectEnablers = objectEnablers;
+        return this;
+    }
+
+    public LeshanClient build() {
```
Comment (jvermillard) [inline leshan-client-cf/src/main/java/org/eclipse/leshan/client/californium/LeshanClientBuilder.java]: javadoc

File: leshan-client-core/src/main/java/org/eclipse/leshan/client/servers/RegistrationEngine.java [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+import org.eclipse.leshan.core.request.RegisterRequest;
+import org.eclipse.leshan.core.request.UpdateRequest;
+import org.eclipse.leshan.core.response.BootstrapResponse;
+import org.eclipse.leshan.core.response.DeregisterResponse;
+import org.eclipse.leshan.core.response.RegisterResponse;
+import org.eclipse.leshan.core.response.UpdateResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RegistrationEngine {
```
Comment (jvermillard) [inline leshan-client-core/src/main/java/org/eclipse/leshan/client/servers/RegistrationEngine.java]: javadoc

File: leshan-client-core/src/main/java/org/eclipse/leshan/client/util/LwM2mId.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+ * 
+ * Contributors:
+ *     Sierra Wireless - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.leshan.client.util;
+
+/**
+ * LWM2M Object/Resource IDs
+ */
+public interface LwM2mId {
```
Comment (jvermillard) [inline leshan-client-core/src/main/java/org/eclipse/leshan/client/util/LwM2mId.java]: leshan-core?

File: leshan-client-core/src/main/java/org/eclipse/leshan/client/servers/RegistrationEngine.java [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+
+    private boolean bootstrap() {
+        ServersInfo serversInfo = ServersInfoExtractor.getInfo(objectEnablers);
+
+        if (serversInfo.bootstrap == null) {
+            LOG.error("Missing info to boostrap the client");
+            return false;
+        }
+
+        if (bootstrapHandler.tryToInitSession(serversInfo.bootstrap)) {
```
Comment (jvermillard) [inline leshan-client-core/src/main/java/org/eclipse/leshan/client/servers/RegistrationEngine.java]: try/catch/finaly

**Comment by msangoi** [comment]
:+1: for merging

**Pull request merged**